docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,554 @@
+          name: release-name
+          namespace: release-namespace 
+      backend: 3ef2e4ac-9070-4093-a3ce-142139fd4a16 # helm-release backend - resolved UUID based on the injected TypeInstance
+      ```
+
+## Workflow syntax - Update
+
+Similarly as with create, Content Developer specifies in the workflow:
+
+```yaml
+capact-updateTypeInstances:
+- name: testUpdate
+  from: update
+```
+
+where the `update` Argo artifact can contain `value` and / or `additionalParameters`.
+
+As backend is immutable, we don't need to provide any additional syntax around TypeInstance update
+
+## Storage Backend service implementation
+
+1. The registered service will be called by Capact Hub and needs to handle following methods:
+
+    TODO: Pseudocode, change to actual HTTP requests / ProtoBuf definition
+
+    ```go
+    // TypeInstance ResourceVersion value
+    getValue(typeInstanceID, additionalParameters?, resourceVersion): (value, error)
+
+    onCreate(typeInstanceID, additionalParameters?, value?): (additionalParameters?, error)
+
+    onUpdate(
+      typeInstanceID,
+      old: {additionalParameters?, resourceVersion},
+      new: {additionalParameters?, resourceVersion},
+      ): (additionalParameters?, error)
+
+    onDelete(typeInstanceID, additionalParameters?) error
+
+    // TypeInstance locking 
+    lockedBy(typeInstanceID, additionalParameters?) (string, error)
+    onLock(typeInstanceID, additionalParameters?) error
+    onUnlock(typeInstanceID, additionalParameters?) error
+    ```
+
+  An implementation of such service may vary between two use cases:
+
+  1. CRUD operations on output TypeInstance actually manages external resource (e.g. Vault) -> onCreate, onUpdate, and onDelete actually creates, updates and deletes a given resource.
+  1. output TypeInstance represents external resources managed in different way (e.g. via Capact actions - like Helm Runner). IMO we shouldn't move actual Helm release installation to TypeInstance "constructor").
+
+      - The service can also implement watch for external resources (e.g. Kubernetes secrets) and call `createTypeInstances` and `deleteTypeInstances` Hub mutations. We may provide Go framework to speed up such development, similarly as we have with Runner concept.
+
+1. The service could be implemented using one of the following solutions (but it is not limited to):
+
+  - [Dapr secrets](https://docs.dapr.io/developing-applications/building-blocks/secrets/secrets-overview/)
+  - [Kubernetes external secrets](https://github.com/external-secrets/kubernetes-external-secrets)
+  - [vault-k8s](https://github.com/hashicorp/vault-k8s)
+  - [db](https://upper.io/v4/getting-started/)
+
+## GraphQL API
+
+The new GraphQL API can be used both on CLI and UI.
+
+To list all available StorageBackends in Hub:
+
+```graphql
+query {
+   types(filter: { pathPattern: "cap.core.hub.storage.*" }) {
```

Comment:
backend can be only under `cap.core.*` or you mean `To list all available (core) StorageBackends in Hub:`?

Reply:
nope, all Storage Backends are attached to the `cap.core.hub.storage` node, similarly what we had in mind when it comes to `platform`: https://capact.io/docs/feature/type-features#additional-references-to-parent-nodes